### PR TITLE
3203 Forest area as proportion of total land area SDG - should be estimated bases on the latest available NDP after 2020

### LINKED ISSUE #3203: Forest area as proportion of total land area SDG - should be estimated bases on the latest available NDP after 2020
@minotogna 

**Describe the bug**
 **Forest area as proportion of total land area** SDG indicator uses the linear trend even though there is a new NDP after 2020.

**To Reproduce**
See Australia(screenshots below)

**Expected behavior**
If the latest data of NDP after 2020 is available, the **Forest area as proportion of total land area** should be estimated based on the latest data of NDP **not the 2020 data**

**Screenshots**
image
image

### PR DESCRIPTION
**Pull request by sorja**
The calculation formula became a bit more complicated when we must know the latest year, added a new file

- Introduced `expressionEvaluator/functuions/calculations/`

Todo:
- Calculation: Verify and understand the logic for all different cases

__Developer note: After a bit of slight annoyance that record format was not supported in calculations (is it..? probably not?) opened an issue to migrate from arena-core and add support for record format (plugin for jsep)__
https://github.com/openforis/fra-platform/issues/3426

__Developer note2:__

<details>
<summary>This is the old calculation formula as reference (year 2023)</summary>

```js
extentOfForest.forestArea['2023'] ? extentOfForest.forestArea['2023'] / extentOfForest.totalLandArea['2023'] * 100 : ((extentOfForest.forestArea['2020'] / extentOfForest.totalLandArea['2020'] * 100 > 0 && extentOfForest.forestArea['2025'] / extentOfForest.totalLandArea['2025'] * 100) > 0 ? (extentOfForest.forestArea['2020'] / extentOfForest.totalLandArea['2020'] * 100 + ((extentOfForest.forestArea['2025'] / extentOfForest.totalLandArea['2025'] * 100 - extentOfForest.forestArea['2020'] / extentOfForest.totalLandArea['2020'] * 100) / 5 * 3)) : null)
```
</details>

### CODE REVIEW

File: src/test/migrations/steps/20240122134404-step-sdg-1511-fix-calculation.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  // Update cache
+  await AssessmentController.generateMetaCache(client)
+  await AssessmentController.generateMetadataCache({ assessment }, client)
+
+  // Update calculated variables
+  const sectionName = 'sustainableDevelopment'
+  const tableName = 'sustainableDevelopment15_1_1'
+  const variableName = 'forestAreaProportionLandArea2015'
+  const updateCalculatedVariableProps = { assessment, cycle, sectionName, tableName, variableName }
+  await updateCalculatedVariable(updateCalculatedVariableProps, client)
```
Comment (sorja) [inline src/test/migrations/steps/20240122134404-step-sdg-1511-fix-calculation.ts]: For some reason this throws :

```
error: Cannot read properties of undefined (reading 'calculations')
```

File: src/test/migrations/steps/20240122134404-step-sdg-1511-fix-calculation.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  // Update cache
+  await AssessmentController.generateMetaCache(client)
+  await AssessmentController.generateMetadataCache({ assessment }, client)
+
+  // Update calculated variables
+  const sectionName = 'sustainableDevelopment'
+  const tableName = 'sustainableDevelopment15_1_1'
+  const variableName = 'forestAreaProportionLandArea2015'
+  const updateCalculatedVariableProps = { assessment, cycle, sectionName, tableName, variableName }
```
Comment (minotogna) [inline src/test/migrations/steps/20240122134404-step-sdg-1511-fix-calculation.ts]: are u sure passing the assessment fetched before updating caches (line 67/68), the updateCalcuatedVariables reads the correct formula? I had to refetch the assessment in some of my previous pr.
can u please verify ?

Comment (sorja) [reply src/test/migrations/steps/20240122134404-step-sdg-1511-fix-calculation.ts]: It seems I was using the deprecated updateCalculatedVariable, I will change it to use the new one

Comment (minotogna) [reply src/test/migrations/steps/20240122134404-step-sdg-1511-fix-calculation.ts]: ok, can u first fix the repository edit global access though ?

Comment (sorja) [reply src/test/migrations/steps/20240122134404-step-sdg-1511-fix-calculation.ts]: sure

**Review by minotogna — APPROVED** [review]

**Pull request merged**